packages/react-storage/src/components/StorageBrowser/Views/LocationsView/Controls/DataTable.tsx

```diff
@@ -0,0 +1,173 @@
+                : 'sort-indeterminate'
+            }
+          />
+        </ButtonElement>
+      ),
+      key: 'th-type',
+      className: `${TABLE_HEADER_CLASS_NAME} ${TABLE_HEADER_CLASS_NAME}--type`,
+      onClick: () => onTableHeaderClick('type'),
+      'aria-sort': selection === 'type' ? direction : 'none',
+    },
+    {
+      children: (
+        <ButtonElement
+          variant="sort"
+          className={TABLE_HEADER_BUTTON_CLASS_NAME}
+        >
+          Permission
+          <IconElement
+            variant={
+              selection === 'permission' && direction !== 'none'
+                ? `sort-${direction}`
+                : 'sort-indeterminate'
+            }
+          />
+        </ButtonElement>
+      ),
+      key: 'th-permission',
+      className: `${TABLE_HEADER_CLASS_NAME} ${TABLE_HEADER_CLASS_NAME}--type`,
+      onClick: () => onTableHeaderClick('permission'),
+      'aria-sort': selection === 'permission' ? direction : 'none',
+    },
+  ];
+
+  const compareFn = LocationsViewColumnSortMap[selection];
+
+  if (direction === 'ascending') {
+    data.sort((a, b) => compareFn(a[selection], b[selection]));
+  } else {
+    data.sort((a, b) => compareFn(b[selection], a[selection]));
+  }
+
+  const rows = data.map((location, index) => [
+    {
+      key: `td-scope-${index}`,
+      children: (
+        <ButtonElement
+          className={TABLE_DATA_BUTTON_CLASS}
+          onClick={() => onLocationClick(location)}
+          variant="table-data"
+        >
+          {location.scope}
+        </ButtonElement>
+      ),
+    },
+    { key: `td-type-${index}`, children: location.type },
+    { key: `td-permission-${index}`, children: location.permission },
+  ]);
+
+  return { columns, rows };
+};
+
+export function LocationsViewTableControl(): React.JSX.Element | null {
+  const [{ data, isLoading }] = useLocationsData();
+
+  const [, handleUpdateState] = useControl({ type: 'NAVIGATE' });
+
+  const [sortState, setSortState] = React.useState<
+    SortState<LocationAccess<Permission>>
+  >({
+    selection: 'scope',
+    direction: 'ascending',
+  });
+
+  const locationsData = React.useMemo(
+    () =>
+      getLocationsData({
+        data: data.result,
+        sortState,
+        onLocationClick: (payload) => {
+          handleUpdateState({
+            type: 'ACCESS_LOCATION',
+            location: payload,
+          });
+        },
+        onTableHeaderClick: (payload: keyof LocationAccess<Permission>) => {
+          setSortState((prevState) => ({
+            selection: payload,
+            direction:
+              prevState.direction === 'ascending' ? 'descending' : 'ascending',
+          }));
+        },
+      }),
+    [data.result, handleUpdateState, sortState]
+  );
+
+  const hasLocations = !!data.result?.length;
+  const shouldRenderLocations = hasLocations && !isLoading;
+
+  return shouldRenderLocations ? <DataTable data={locationsData} /> : null;
```

Comment:
`DataTableControl` should not be responsible for conditional rendering, can remove the logic here